drivers/clock_control/clock_control_bl60x.c

```diff
@@ -0,0 +1,923 @@
+	clock_control_bl60x_clock_settle();
+
+	return 0;
+}
+
+static void clock_control_bl60x_set_machine_timer_clock_enable(uint32_t enable)
+{
+	uint32_t tmp;
+
+	if (enable > 1) {
+		enable = 1;
+	}
+	tmp = sys_read32(GLB_BASE + GLB_CPU_CLK_CFG_OFFSET);
+	tmp = (tmp & GLB_CPU_RTC_EN_UMSK) | (enable << GLB_CPU_RTC_EN_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CPU_CLK_CFG_OFFSET);
+}
+
+/* clock:
+ * 0: BCLK
+ * 1: 32Khz Oscillator (RC32*K*)
+ */
+static void clock_control_bl60x_set_machine_timer_clock(uint32_t enable, uint32_t clock,
+							uint32_t divider)
+{
+	uint32_t tmp;
+
+	if (divider > 0x1FFFF) {
+		divider = 0x1FFFF;
+	}
+	if (clock > 1) {
+		clock = 1;
+	}
+
+	/* disable first, then set div */
+	clock_control_bl60x_set_machine_timer_clock_enable(0);
+
+	tmp = sys_read32(GLB_BASE + GLB_CPU_CLK_CFG_OFFSET);
+	tmp = (tmp & GLB_CPU_RTC_SEL_UMSK) | (clock << GLB_CPU_RTC_SEL_POS);
+	tmp = (tmp & GLB_CPU_RTC_DIV_UMSK) | (divider << GLB_CPU_RTC_DIV_POS);
+	sys_write32(tmp, GLB_BASE + GLB_CPU_CLK_CFG_OFFSET);
+
+	clock_control_bl60x_set_machine_timer_clock_enable(enable);
+}
+
+static void clock_control_bl60x_deinit_pll(void)
+{
+	uint32_t tmp;
+
+	/* PLL Off */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_PU_CLKPLL_SFREG_UMSK) | ((uint32_t)(0) << PDS_PU_CLKPLL_SFREG_POS);
+	tmp = (tmp & PDS_PU_CLKPLL_UMSK) | ((uint32_t)(0) << PDS_PU_CLKPLL_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	/* needs 2 steps ? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_PU_CP_UMSK) | ((uint32_t)(0) << PDS_CLKPLL_PU_CP_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_PFD_UMSK) | ((uint32_t)(0) << PDS_CLKPLL_PU_PFD_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_FBDV_UMSK) | ((uint32_t)(0) << PDS_CLKPLL_PU_FBDV_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_POSTDIV_UMSK) | ((uint32_t)(0) << PDS_CLKPLL_PU_POSTDIV_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+}
+
+/* RC32M : 0
+ * XTAL : 1
+ */
+static void clock_control_bl60x_set_pll_source(uint32_t source)
+{
+	uint32_t tmp;
+
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+	if (source > 0) {
+		tmp = (tmp & PDS_CLKPLL_REFCLK_SEL_UMSK) |
+		      ((uint32_t)(1) << PDS_CLKPLL_REFCLK_SEL_POS);
+		tmp = (tmp & PDS_CLKPLL_XTAL_RC32M_SEL_UMSK) |
+		      ((uint32_t)(0) << PDS_CLKPLL_XTAL_RC32M_SEL_POS);
+	} else {
+		tmp = (tmp & PDS_CLKPLL_REFCLK_SEL_UMSK) |
+		      ((uint32_t)(0) << PDS_CLKPLL_REFCLK_SEL_POS);
+		tmp = (tmp & PDS_CLKPLL_XTAL_RC32M_SEL_UMSK) |
+		      ((uint32_t)(1) << PDS_CLKPLL_XTAL_RC32M_SEL_POS);
+	}
+	sys_write32(tmp, PDS_BASE + PDS_CLKPLL_TOP_CTRL_OFFSET);
+}
+
+static void clock_control_bl60x_init_pll(enum bl60x_clkid source, uint32_t crystal_frequency)
+{
+	uint32_t tmp;
+	uint32_t old_rootclk;
+
+	old_rootclk = clock_control_bl60x_get_root_clock();
+
+	/* security RC32M */
+	if (old_rootclk > 1) {
+		clock_control_bl60x_set_root_clock(0);
+	}
+
+	if (clock_control_bl60x_crystal_to_id(crystal_frequency) < 0) {
```

Comment:
What does it mean when this function returns -1?
If `crystal_freq` can only be one of this 5 values, we can probably map them with macro magic and trigger a compilation issue instead of checking in runtime

Reply:
done